core/src/OpenACC/Kokkos_OpenACC_Team.hpp

```diff
@@ -0,0 +1,607 @@
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_TEAM_HPP
+#define KOKKOS_OPENACC_TEAM_HPP
+
+#include <openacc.h>
+#include <impl/Kokkos_Traits.hpp>
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <Kokkos_ExecPolicy.hpp>
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Impl {
+
+class OpenACCTeamMember {
+ public:
+  constexpr static int TEAM_REDUCE_SIZE = 512;
+  // FIXME_OPENACC: default-team-size macros are temporarily used for
+  // team_size_max and team_size_recommended APIs
+  constexpr static int DEFAULT_TEAM_SIZE_MAX = 512;
+  constexpr static int DEFAULT_TEAM_SIZE_REC = 128;
+
+  using execution_space      = Kokkos::Experimental::OpenACC;
+  using scratch_memory_space = execution_space::scratch_memory_space;
+
+  scratch_memory_space m_team_shared;
+  int m_team_scratch_size[2];
+  int m_team_rank;
+  int m_team_size;
+  int m_league_rank;
+  int m_league_size;
+  int m_vector_length;
+
+ public:
+  KOKKOS_FUNCTION
+  const execution_space::scratch_memory_space& team_shmem() const {
+    return m_team_shared.set_team_thread_mode(0, 1, 0);
+  }
+
+  KOKKOS_FUNCTION
+  const execution_space::scratch_memory_space& team_scratch(int level) const {
+    return m_team_shared.set_team_thread_mode(level, 1,
+                                              m_team_scratch_size[level]);
+  }
+
+  KOKKOS_FUNCTION
+  const execution_space::scratch_memory_space& thread_scratch(int level) const {
+    return m_team_shared.set_team_thread_mode(level, team_size(), team_rank());
+  }
+
+  KOKKOS_FUNCTION int league_rank() const { return m_league_rank; }
+  KOKKOS_FUNCTION int league_size() const { return m_league_size; }
+  KOKKOS_FUNCTION int team_rank() const {
+#ifdef KOKKOS_COMPILER_NVHPC
+    return __pgi_vectoridx();
+#else
+    return m_team_rank;
```

Comment:
Why does `team_rank()` not return `m_team_rank`, using NVHPC?

Reply:
Move `__pgi_vectoridx()` call, which is used to set `m_team_rank` variable, into the OpenACCTeamMember constructor.